library-udf/src/main/java/org/apache/iotdb/library/frequency/UDTFIDWT.java

```diff
@@ -0,0 +1,97 @@
+package org.apache.iotdb.library.frequency;
+
+import org.apache.iotdb.db.query.udf.api.UDTF;
+import org.apache.iotdb.db.query.udf.api.access.Row;
+import org.apache.iotdb.db.query.udf.api.collector.PointCollector;
+import org.apache.iotdb.db.query.udf.api.customizer.config.UDTFConfigurations;
+import org.apache.iotdb.db.query.udf.api.customizer.parameter.UDFParameterValidator;
+import org.apache.iotdb.db.query.udf.api.customizer.parameter.UDFParameters;
+import org.apache.iotdb.db.query.udf.api.customizer.strategy.RowByRowAccessStrategy;
+import org.apache.iotdb.library.frequency.util.DWTUtil;
+import org.apache.iotdb.library.util.Util;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+
+import java.util.ArrayList;
+
+/**
+ * This function calculates inverse discrete wavelet transform of an input series, which is composed
+ * of multiple layers. Input series must be an integer exponent of 2. Input series is treated as
+ * sampled in equal distances.
+ */
+public class UDTFIDWT implements UDTF {
+  private ArrayList<Long> timestamp = new ArrayList<>();
+  private ArrayList<Double> value = new ArrayList<>();
+  private String s;
+  private String method;
+  private int layer;
+
+  @Override
+  public void validate(UDFParameterValidator validator) throws Exception {
+    validator
+        .validateInputSeriesDataType(
+            0, TSDataType.DOUBLE, TSDataType.FLOAT, TSDataType.INT32, TSDataType.INT64)
+        .validate(
+            x ->
+                ((String) x).equalsIgnoreCase("Haar")
+                    || ((String) x).equalsIgnoreCase("DB2")
+                    || ((String) x).equalsIgnoreCase("DB4")
+                    || ((String) x).equalsIgnoreCase("DB6")
+                    || ((String) x).equalsIgnoreCase("DB8")
+                    || ((String) x).equalsIgnoreCase(""),
+            "Method not supported, please input coefficient and leave method blank.",
+            validator.getParameters().getStringOrDefault("method", ""))
+        .validate(
+            x -> (int) x > 0,
+            "layer has to be a positive integer.",
+            validator.getParameters().getIntOrDefault("layer", 1));
+  }
+
+  @Override
+  public void beforeStart(UDFParameters parameters, UDTFConfigurations configurations)
+      throws Exception {
+    configurations
+        .setAccessStrategy(new RowByRowAccessStrategy())
+        .setOutputDataType(TSDataType.DOUBLE);
+    s = parameters.getString("coef");
+    method = parameters.getStringOrDefault("method", "");
+    layer = parameters.getIntOrDefault("layer", 1);
+    timestamp.clear();
+    value.clear();
+  }
+
+  @Override
+  public void transform(Row row, PointCollector pointCollector) throws Exception {
+    timestamp.add(row.getTime());
+    value.add(Util.getValueAsDouble(row));
+  }
+
+  @Override
+  public void terminate(PointCollector pointCollector) throws Exception {
+    if (!s.equals("") || !method.equals("")) { // When user offers at least one parameter
```

Comment:
Good Job.Thanks for your contribution.This adds a lot of udf to us.
I have a littile..
if s ==null, we will...

Reply:
By changing `s=parameters.getString()` to `s=parameters.getStringOrDefault()` this problem is fixed.